### PR TITLE
[POAE7-2689] simple non-groupby agg translator (support sum)

### PR DESCRIPTION
**Pull request by YantingTao1315**
### What changes were proposed in this pull request?
+ support simple agg translator (sum)
+ add related logic in paser
+ add related runtime function
+ add related test
+ add initializer in context buffer

### Why are the changes needed?
+ agg translator implement

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UTs

### Which label does this PR belong to?
INFRA

### CODE REVIEW

File: cider/exec/nextgen/operators/OperatorRuntimeFunctions.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+#ifndef NEXTEGN_CIDER_FUNCTION_RUNTIME_FUNCTIONS_H
+#define NEXTEGN_CIDER_FUNCTION_RUNTIME_FUNCTIONS_H
+
+#include "exec/nextgen/context/RuntimeContext.h"
+#include "type/data/funcannotations.h"
+
+/********************* Simple Aggregation Functions For Nextgen
+ * *****************************/
+#define DEF_NEXTEGN_CIDER_CIDER_SIMPLE_AGG_INT(width, aggname, aggfunc)   \
```
Comment (jikunshang) [inline cider/exec/nextgen/operators/OperatorRuntimeFunctions.h:30]: DEF_NEXTEGN_CIDER_SIMPLE_AGG_INT

File: cider/exec/nextgen/operators/AggNode.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  AggExprsInfo(SQLTypes st, SQLAgg at, int32_t so, int32_t bs);
+};
+
+using AggExprsInfoVector = std::vector<AggExprsInfo>;
+
+class AggNode : public OpNode {
+ public:
+  AggNode(ExprPtrVector&& groupby_exprs, ExprPtrVector&& output_exprs)
+      : OpNode("AggNode", std::move(output_exprs), JITExprValueType::ROW)
+      , groupby_exprs_(groupby_exprs) {}
```
Comment (jikunshang) [inline cider/exec/nextgen/operators/AggNode.h:45]: std::move(groupby_exprs) ?

File: cider/exec/nextgen/operators/AggNode.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+  AggExprsInfoVector exprs_info;
+  initialExpersInfo(exprs_info, exprs);
+
+  std::vector<int8_t> origin_value(exprs_info.back().start_offset +
+                                   exprs_info.back().byte_size);
+  initialOriginValue(origin_value, exprs_info);
+
+  // Groupby
+  // TODO(Yanting): support group-by
+  auto agg_node = dynamic_cast<AggNode*>(node_.get());
```
[the remark this answers is not in the document]
Comment (YantingTao1315) [reply cider/exec/nextgen/operators/AggNode.cpp:161]: groupby exprs is  disabled in parser
image
this path will not be test, and groupby related test will be add in next pr(enable groupby)

File: cider/exec/nextgen/operators/AggNode.cpp [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+TranslatorPtr AggNode::toTranslator(const TranslatorPtr& succ) {
+  return createOpTranslator<AggTranslator>(shared_from_this(), succ);
+}
+
+void AggTranslator::consume(context::CodegenContext& context) {
+  codegen(context);
+}
+
+void initialExpersInfo(AggExprsInfoVector& infos, ExprPtrVector& exprs) {
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/operators/AggNode.cpp:94]: please move the output argument after the input argument. i.e `initialExpersInfo(ExprPtrVector& exprs, AggExprsInfoVector& infos)`

File: cider/exec/nextgen/operators/AggNode.h [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  AggNode(ExprPtrVector&& groupby_exprs, ExprPtrVector&& output_exprs)
+      : OpNode("AggNode", std::move(output_exprs), JITExprValueType::ROW)
+      , groupby_exprs_(std::move(groupby_exprs)) {}
+
+  AggNode(const ExprPtrVector& groupby_exprs, const ExprPtrVector& output_exprs)
+      : OpNode("AggNode", output_exprs, JITExprValueType::ROW)
+      , groupby_exprs_(std::move(groupby_exprs)) {}
+
+  std::pair<JITExprValueType, ExprPtrVector&> getGroupByExprs() {
+    return {output_type_, groupby_exprs_};
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/operators/AggNode.h:52]: `output_type_` is used for `output_exprs_`, not for `groupby_exprs_`

File: cider/exec/nextgen/parsers/RelAlgEUParser.cpp [hunk 6]
```diff
@@ -138,11 +135,28 @@ OpPipeline toOpPipeline(const RelAlgExecutionUnit& eu) {
...
+  ExprPtrVector aggs;
+  ExprPtrVector groupbys;
+
   for (auto& targets_expr : eu.shared_target_exprs) {
-    projs.push_back(targets_expr);
+    if (targets_expr->get_contains_agg()) {
+      aggs.push_back(targets_expr);
+    } else {
+      projs.push_back(targets_expr);
+    }
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/parsers/RelAlgEUParser.cpp:146]: how do we keep the order of the output column in `select a, sum(*), b, count(*) .. group by a, b`

File: cider/exec/nextgen/parsers/RelAlgEUParser.cpp [hunk 7]
```diff
@@ -138,11 +135,28 @@ OpPipeline toOpPipeline(const RelAlgExecutionUnit& eu) {
...
   ops.emplace_back(createOpNode<operators::ProjectNode>(projs));
 
+  if (!eu.groupby_exprs.empty() && eu.groupby_exprs.front()) {
+    for (auto& groupby_expr : eu.groupby_exprs) {
+      groupbys.push_back(groupby_expr);
+    }
+  }
+
+  if (!groupbys.empty() || !aggs.empty()) {
+    ops.emplace_back(createOpNode<operators::AggNode>(groupbys, aggs));
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/parsers/RelAlgEUParser.cpp:157]: may be we need insert AggNode before ProjectNode

File: cider/tests/nextgen/compiler/AggTest.cpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+    auto buffer = reinterpret_cast<cider::exec::nextgen::context::Buffer*>(
+        runtime_ctx->getContextItem(0));
+    auto under_buffer = reinterpret_cast<int64_t*>(buffer->getBuffer());
+    EXPECT_EQ(under_buffer[0], 23);
+    EXPECT_EQ(under_buffer[1], 1293);
+  }
+
+ private:
+  std::string create_ddl_ = "CREATE TABLE test(a BIGINT, b BIGINT);";
```
Comment (Ma-Jian1) [inline cider/tests/nextgen/compiler/AggTest.cpp:102]: add a test for a, b not the same type?

File: cider/exec/nextgen/operators/AggNode.cpp [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+  auto buffer_initializer = [origin_value](context::Buffer* buf) {
+    auto raw_buf = buf->getBuffer();
+    memcpy(raw_buf, origin_value.data(), buf->getCapacity());
+  };
+  auto buffer =
+      context.registerBuffer(origin_value.size(), "output_buffer", buffer_initializer);
+
+  int32_t current_expr_idx = 0;
+  for (const auto& expr : exprs) {
+    auto agg_expr = dynamic_cast<const Analyzer::AggExpr*>(expr.get());
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/operators/AggNode.cpp:178]: please move these into AggExpr::codegen

File: cider/exec/nextgen/operators/AggNode.cpp (hunk 4 above)
Comment (bigPYJ1151) [reply cider/exec/nextgen/operators/AggNode.cpp:94]: Why not make ```infos``` as a return value? Meanwhile I think named as ```initExprsInfo``` or ```initializeExprsInfo``` is better.

File: cider/exec/nextgen/operators/AggNode.cpp [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        break;
+    }
+    AggExprsInfo info = AggExprsInfo(
+        agg_expr->get_type_info().get_type(), agg_expr->get_aggtype(), start_addr, size);
+    start_addr += size;
+    infos.push_back(info);
+  }
+}
+
+void initialOriginValue(std::vector<int8_t>& origin_vextor,
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply cider/exec/nextgen/operators/AggNode.cpp:118]: As before, why not make ```origin_vextor``` as a return value? ```initial``` should be replaced as ```init``` or ```initialize```.

File: cider/exec/nextgen/operators/AggNode.cpp [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+    auto raw_buf = buf->getBuffer();
+    memcpy(raw_buf, origin_value.data(), buf->getCapacity());
+  };
+  auto buffer =
+      context.registerBuffer(origin_value.size(), "output_buffer", buffer_initializer);
+
+  int32_t current_expr_idx = 0;
+  for (const auto& expr : exprs) {
+    auto agg_expr = dynamic_cast<const Analyzer::AggExpr*>(expr.get());
+    auto value = agg_expr->get_arg()->get_expr_value();
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply cider/exec/nextgen/operators/AggNode.cpp:179]: Perhaps you could refactor as:
```c++
utils::FixSizeJITExprValue values(agg_expr->get_arg()->get_expr_value());
```

File: cider/exec/nextgen/operators/AggNode.h [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+#ifndef NEXTGEN_OPERATORS_AGGNODE_H
+#define NEXTGEN_OPERATORS_AGGNODE_H
+
+#include "exec/nextgen/operators/OpNode.h"
+
+namespace cider::exec::nextgen::operators {
+struct AggExprsInfo {
+  jitlib::JITTypeTag jit_value_type;
+  SQLTypes sql_types;
```
Comment (bigPYJ1151) [inline cider/exec/nextgen/operators/AggNode.h:30]: Seems this field is no use.

File: cider/exec/nextgen/operators/AggNode.cpp [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+
+void AggTranslator::consume(context::CodegenContext& context) {
+  codegen(context);
+}
+
+void initialExpersInfo(AggExprsInfoVector& infos, ExprPtrVector& exprs) {
+  int32_t start_addr = 0;
+  for (const auto& expr : exprs) {
+    int32_t size = 0;
+    auto agg_expr = dynamic_cast<const Analyzer::AggExpr*>(expr.get());
```
[the remark this answers is not in the document]
Comment (YantingTao1315) [reply cider/exec/nextgen/operators/AggNode.cpp:98]: exprs have been classified before

File: cider/exec/nextgen/operators/AggNode.cpp [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+  int8_t* raw_memory = origin_vextor.data();
+  for (const auto& info : exprs_info) {
+    switch (info.agg_type) {
+      case SQLAgg::kSUM:
+      case SQLAgg::kCOUNT: {
+        switch (info.byte_size) {
+          case 1:
+          // how to change type
+          case 8: {
+            auto cast_memory = reinterpret_cast<int64_t*>(raw_memory + info.start_offset);
```
[the remark this answers is not in the document]
Comment (YantingTao1315) [reply cider/exec/nextgen/operators/AggNode.cpp:129]: use offset indtead

File: cider/exec/nextgen/parsers/RelAlgEUParser.cpp (hunk 6 above)
Comment (YantingTao1315) [reply cider/exec/nextgen/parsers/RelAlgEUParser.cpp:146]: this will fix in pr which supports groupby

**Review by Ma-Jian1 — DISMISSED** [review]

**Review by bigPYJ1151 — APPROVED** [review]

**Review by Ma-Jian1 — APPROVED** [review]

**Review by harborn — APPROVED** [review]

**Pull request merged**